envoy-sdk-test/src/host/simulate.rs

```diff
@@ -33,3 +33,33 @@ pub fn get_buffer_bytes(buf: &[u8], offset: usize, max_size: usize) -> host::Res
     }
     Ok(ByteString::default())
 }
+
+/// Mutates buffer similarly to `Proxy Wasm` inside Envoy.
+pub fn set_buffer_bytes(
+    buf: &mut Vec<u8>,
+    start: usize,
+    length: usize,
+    data: &[u8],
+) -> host::Result<()> {
+    // implementation based on `envoyproxy/envoy-wasm`
+
+    if start == 0 {
+        if length == 0 {
+            let tail: Vec<u8> = buf.drain(..).collect();
+            buf.extend(data);
+            buf.extend(tail);
+            Ok(())
+        } else if length >= buf.len() {
+            buf.truncate(0);
+            buf.extend(data);
+            Ok(())
+        } else {
+            Err(format_err!("Status::BadArgument"))
+        }
+    } else if start >= buf.len() {
+        buf.extend(data);
+        Ok(())
+    } else {
+        Err(format_err!("Status::BadArgument"))
+    }
```

Comment:
I would write it like this:

```suggestion
    let chunk = if start == 0 {length} else {start};
    if chunk != 0 && chunk < buf.len() {
        Err(format_err!("Status::BadArgument"))
    } else {
        let head: Vec<u8> = buf.drain(..min(start, buf.len())).collect();
        buf.drain(..min(length, buf.len()));
        let tail: Vec<u8> = buf.drain(..).collect();

        buf.extend(head);
        buf.extend(data);
        buf.extend(tail);

        Ok(())
    }
```
This might make it clearer what the function does and doesn't do and what the parameters mean. There might be a cleaner way to express this but I don't claim to be an expert on rust arrays. Either way this isn't that big a deal so you're welcome to just leave it the way it is.

Reply:
Ah, this is the only code where I'm trying to do idiomatic Rust but rather reproduce Envoy/Proxy Wasm logic "as is" (for the sake of easier updates later on).

I like the logic in the `else` block, but logic around `chunk` is too convoluted for me (again, for the purpose of constantly comparing it against original implementation in Envoy).

I will think how to incorporate this suggestion.

Reply:
The chunk block could just as easily be two if statements. I was trying to come up with a way to make the general case more clear (by having one case) and then expressing the unsupported cases. My review comments here aren't especially enlightened, in that structural feedback is more important than what I've provided so far, so feel free to think of this as "I might've done this differently" and simply move on.